.gitbook/contents/developer/features/aa-basics/hybrid-compute.md

```diff
@@ -0,0 +1,549 @@
+
+    p = ethabi.encode([
+        'address',
+        'uint256',
+        'bytes32',
+        'bytes32',
+        'uint256',
+        'uint256',
+        'uint256',
+        'uint256',
+        'uint256',
+        'bytes32',
+    ], [
+        HybridAcctAddr,
+        req['opNonce'],
+        Web3.keccak(Web3.to_bytes(hexstr='0x')),  # initCode
+        Web3.keccak(Web3.to_bytes(hexstr=p_enc2)),
+        callGas,
+        Web3.to_int(hexstr=limits['verificationGasLimit']),
+        Web3.to_int(hexstr=limits['preVerificationGas']),
+        0,  # maxFeePerGas
+        0,  # maxPriorityFeePerGas
+        Web3.keccak(Web3.to_bytes(hexstr='0x')),  # paymasterANdData
+    ])
+
+    ooHash = Web3.keccak(ethabi.encode(['bytes32', 'address', 'uint256'], [
+                         Web3.keccak(p), EntryPointAddr, HC_CHAIN]))
+
+    signAcct = eth_account.account.Account.from_key(hc1_key)
+    eMsg = eth_account.messages.encode_defunct(ooHash)
+    sig = signAcct.sign_message(eMsg)
+
+    success = (err_code == 0)
+    print("Method returning success={} response={} signature={}".format(
+        success, Web3.to_hex(resp_payload), Web3.to_hex(sig.signature)))
+    return ({
+        "success": success,
+        "response": Web3.to_hex(resp_payload),
+        "signature": Web3.to_hex(sig.signature)
+    })
+```
+
+With that settled, we have successfuly implemented a function, which can receive a request from the bundler, perform
+some calculation with its payload and return a response.
+
+## Step 2: Setting up a server
+
+With the offchain handler created, the next step is to set up a server to run it. In this example, we will create a
+simple JSON-RPC server using the jsonrpclib library. We define
+
+```python
+from jsonrpclib.SimpleJSONRPCServer import SimpleJSONRPCServer, SimpleJSONRPCRequestHandler
+
+def selector(name):
+    nameHash = Web3.to_hex(Web3.keccak(text=name))
+    return nameHash[2:10]
+
+
+class RequestHandler(SimpleJSONRPCRequestHandler):
+    rpc_paths = ('/', '/hc')
+
+
+def server_loop():
+    server = SimpleJSONRPCServer(
+        ('0.0.0.0', 1234), requestHandler=RequestHandler)
+    server.register_function(offchain_addsub2, selector(
+        "addsub2(uint32,uint32)"))  # ASD97e0d7ba
+
+    server.serve_forever()
+
+
+server_loop()  # Run until killed
+```
+
+Here, we register our function ``offchain_addsub2``. Note the "identifier" passed to register_function. This identifier
+is
+generated by the selector function, which creates a hashed representation of the function signature provided as an
+argument.
+
+### Why is that?
+
+The bundler sends us an JSON-RPC v2 request, containing a ``method`` identifier which is the function selector of the
+desired offchain method, along with
+several standard parameters (NOTE - names are subject to change) as well as a "payload" which contains the ABI-encoded
+request data:
+
+``` JSON
+{
+  "jsonrpc":"2.0",
+  "id":0,
+  "method":"ASD97e0d7ba",
```

Comment:
One question I have is whether it makes sense to implement the JSON-RPC "method" like this, or whether people would prefer to call it something fixed like "eth.offchain" and put the function selector into its own field as part of the request params. Opinions are welcome.

Reply:
yes the more cleartext the better imho.